File: Python_Books/算法图解/grokking_algorithms_COPY/04_quicksort/lua/04_recursive_max.lua
function max(array)
    if next(array) == nil then
        return nil
    elseif #array == 1 then
        return array[1]
    elseif #array == 2 then
        return array[1] > array[2] and array[1] or array[2]
    end
    
    local sub_max = max(table.move(array, 2, #array, 1, {}))
    return array[1] > sub_max and array[1] or sub_max
end

print(max({1, 2, 3, 4})) -- => 4

File: Python_Books/算法图解/grokking_algorithms_COPY/04_quicksort/javascript/04_recursive_max.js
'use strict';

function max(list) {
  if (list.length === 2) {
    return list[0] > list[1] ? list[0] : list[1];
  }
  let sub_max = max(list.slice(1));
  return list[0] > sub_max ? list[0] : sub_max;
}

console.log(max([1, 5, 10, 25, 16, 1])); // 25
